test/distributed/_composable/fsdp/test_fully_shard_compile.py

```diff
@@ -100,6 +135,184 @@ def f(x):
         self.assertEqual(cnt.op_count, 1)
         self.assertEqual(len(cnt.graphs), 1)
 
+    @torch._dynamo.config.patch(inline_inbuilt_nn_modules=True)
+    @torch._functorch.config.patch(recompute_views=True)
+    def _test_traceable_fsdp(
+        self, model_init_fn, input_creation_fn, backend, fullgraph
+    ):
+        n_iter = 10
+
+        def compiler_fn(compiled_autograd_backend):
+            def _fn(gm):
+                # fullgraph=True because graph-break in Compiled Autograd BWD graph is not supported by Traceable FSDP2 yet
+                # (main difficulty comes from queue_callback not working well when BWD has graph break).
+                return torch.compile(
+                    gm, backend=compiled_autograd_backend, fullgraph=True
+                )
+
+            return _fn
+
+        def run_all_iters(model, optim, compiled_autograd_backend=None):
+            torch.manual_seed(42)
+            losses = []
+            for i in range(n_iter):
+                optim.zero_grad(set_to_none=True)
+                inp = input_creation_fn()
+                if compiled_autograd_backend is not None:
+                    maybe_compiled_autograd_ctx = compiled_autograd.enable(
+                        compiler_fn(compiled_autograd_backend)
+                    )
+                else:
+                    maybe_compiled_autograd_ctx = contextlib.nullcontext()
+                with maybe_compiled_autograd_ctx:
+                    out = model(inp)
+                    loss = out.sum()
+                    losses.append(loss.item())
+                    loss.backward()
+                optim.step()
+                del loss
+                del out
+                del inp
+                torch.cuda.synchronize()
+            return losses
+
+        def test_compiled():
+            model, optim = model_init_fn()
+            # FSDP2 does lazy init using 1st run, so run it once to init using eager mode
+            run_all_iters(model, optim, 1)
+
+            model_compiled = torch.compile(model, backend=backend, fullgraph=True)
+            res = run_all_iters(
+                model_compiled, optim, compiled_autograd_backend=backend
+            )
+            cleanup_fsdp(model)
+            del model_compiled
+            del model
+            optim.zero_grad(set_to_none=True)
+            del optim
```

Comment:
For these, I wonder why we need to `del` explicitly? I would expect everything to get GC'd after this function returns.

Reply:
yes we probably don't need to do the `del`s, although the `resize_(0)` might be more fundamental (I can remove all these and investigate again when I get to the memory tests later)

Reply:
Removed all `del`s